### PR TITLE
feat: Extend blueprint interface

### PR DESCRIPTION
**Pull request by kainpets**
Part of #1

### CODE REVIEW

File: src/lib/blueprint.ts [hunk 1]
```diff
@@ -2,6 +2,24 @@ import type { Openapi } from './openapi.js'
...
 export interface Blueprint {
   name: string
+  language: string
+  methodDescription: string
+  versionInformation: string
+  category: string // e.g. Devices, Access Codes, Locks, etc.
+  parameters: Array<{
+    name: string
+    type: string
+    optionality: 'optional' | 'required'
```
Comment (andrii-balitskyi) [inline src/lib/blueprint.ts:12]:
```suggestion
    required: boolean
```

File: src/lib/blueprint.ts [hunk 2]
```diff
@@ -2,6 +2,24 @@ import type { Openapi } from './openapi.js'
 
 export interface Blueprint {
   name: string
+  language: string
+  methodDescription: string
+  versionInformation: string
```
Comment (andrii-balitskyi) [inline src/lib/blueprint.ts:7]: I think `versionInformation` is specific to SDK versioning so probably it shouldn't be part of the blueprint

File: src/lib/blueprint.ts [hunk 3]
```diff
@@ -2,6 +2,24 @@ import type { Openapi } from './openapi.js'
...
+    description: string
+  }>
+  returnValue: {
+    type: string
+    description: string
+  }
+  examples: Array<{
+    codeSnippet: string
+    explanation: string
+  }>
```
Comment (andrii-balitskyi) [inline src/lib/blueprint.ts:22]: Your blueprint type doesn't have a general `routes` or similar property. Instead, method descriptions, parameters, return values, etc., are provided at the top level. Are you proposing to create a blueprint per endpoint? My understanding was that we were including all the routes in the blueprint. Maybe I misunderstood.
@razor-x wdyt?

File: src/lib/openapi.ts [hunk 4]
```diff
@@ -1,5 +1,55 @@
...
 export interface Openapi {
   info: {
     title: string
   }
 }
+
+const extractSchemas = (schemas) => {
+  return Object.entries(schemas).map(([schemaName, schema]) => ({
+    name: schemaName,
+    properties: schema.properties ? Object.keys(schema.properties) : [],
```
Comment (andrii-balitskyi) [inline src/lib/openapi.ts:14]: I think we'll need to more than just property names? Like type, description, required props.

File: src/lib/openapi.ts [hunk 5]
```diff
@@ -1,5 +1,55 @@
...
+  }))
+}
+
+const extractPaths = (paths) => {
+  return Object.entries(paths).map(([path, operations]) => ({
+    path,
+    operations: Object.entries(operations).map(
+      ([method, operationDetails]) => ({
+        method: method.toUpperCase(),
+        summary: operationDetails.summary || '',
```
Comment (andrii-balitskyi) [inline src/lib/openapi.ts:24]: `summary` is the same as `path` so probably we don't need it

File: src/lib/openapi.ts [hunk 6]
```diff
@@ -1,5 +1,55 @@
...
+        operationId: operationDetails.operationId || '',
```
Comment (andrii-balitskyi) [inline src/lib/openapi.ts:25]: What is `operationId` for?

Comment (kainpets) [reply src/lib/openapi.ts:78]: The path once again, removing it

File: src/lib/blueprint.ts (hunk 3 above)
Comment (razor-x) [reply src/lib/blueprint.ts:21]: The Blueprint type is the root type, it needs to contain `routes` or `endpoints` etc.

File: src/lib/openapi.ts [hunk 7]
```diff
@@ -1,5 +1,112 @@
...
+    },
+    schemas,
+    apiPaths,
   }
 }
+
+const extractedOpenapi = extractOpenapi()
+
+const outputFilePath = './extractedOpenapi.json'
+writeFile(outputFilePath, JSON.stringify(extractedOpenapi, null, 2))
```
Comment (razor-x) [inline src/lib/openapi.ts:104]: This file should not implement any I/O. Tests are already setup here: https://github.com/seamapi/blueprint/tree/main/test

File: src/lib/openapi.ts [hunk 8]
```diff
@@ -1,5 +1,112 @@
+/* eslint-disable @typescript-eslint/prefer-nullish-coalescing */
+/* eslint-disable @typescript-eslint/strict-boolean-expressions */
```
Comment (razor-x) [inline src/lib/openapi.ts:2]:
```suggestion
```

File: package.json [hunk 9]
```diff
@@ -69,7 +69,8 @@
   },
   "devDependencies": {
     "@seamapi/types": "1.184.0",
-    "@types/node": "^20.8.10",
+    "@types/eslint": "^8.56.10",
```
Comment (razor-x) [inline package.json:72]: Why was this added? And why did the other dependencies need explicit updates, was something causing an issue?

Comment (kainpets) [reply package.json]: Didn't notice these were already present.

File: src/lib/blueprint.ts [hunk 10]
```diff
@@ -1,7 +1,33 @@
...
 export interface Blueprint {
   name: string
+  routes: {
+    path: string
+    method: string
+    methodDescription: string
+    category: string
+    parameters: Parameter[]
+    returnValue: ReturnValue
+    examples: Example[]
```
Comment (razor-x) [inline src/lib/blueprint.ts:29]: Where would these be used? I don't think examples are in scope, wouldn't any examples be written in the api-docs repo?

File: src/lib/blueprint.ts [hunk 11]
```diff
@@ -1,7 +1,33 @@
...
+  codeSnippet: string
+  explanation: string
+}
+
 export interface Blueprint {
   name: string
+  routes: {
+    path: string
+    method: string
+    methodDescription: string
```
Comment (razor-x) [inline src/lib/blueprint.ts:25]: What is this methodDescription?

File: src/lib/blueprint.ts [hunk 12]
```diff
@@ -1,7 +1,33 @@
...
+    category: string
+    parameters: Parameter[]
+    returnValue: ReturnValue
```
Comment (razor-x) [inline src/lib/blueprint.ts:28]:
```suggestion
    response: Response
```

File: src/lib/blueprint.ts [hunk 13]
```diff
@@ -1,7 +1,33 @@
 import type { Openapi } from './openapi.js'
 
+interface Parameter {
+  name: string
+  type: string
```
Comment (razor-x) [inline src/lib/blueprint.ts:5]: We need to define this as an enum. What types are we supporting? How would this be used? I suggest we actually remove this until we identify what we need.
```suggestion
```

File: src/lib/blueprint.ts [hunk 14]
```diff
@@ -1,7 +1,33 @@
...
+  required: boolean
```
Comment (razor-x) [inline src/lib/blueprint.ts:6]:
```suggestion
  isRequired: boolean
```

File: src/lib/blueprint.ts [hunk 15]
```diff
@@ -1,7 +1,33 @@
...
+  description: string
+}
+
+interface ReturnValue {
+  type: string
```
Comment (razor-x) [inline src/lib/blueprint.ts:11]:
```suggestion
```

File: src/lib/blueprint.ts (hunk 10 above)
Comment (kainpets) [reply src/lib/blueprint.ts:29]: I thought we're thinking about things that would be nice to have but you're right, they won't be here.

File: src/lib/blueprint.ts [hunk 16]
```diff
@@ -1,7 +1,33 @@
...
+  explanation: string
+}
+
 export interface Blueprint {
   name: string
+  routes: {
+    path: string
+    method: string
+    methodDescription: string
+    category: string
```
[the remark this answers is not in the document]
Comment (kainpets) [reply src/lib/blueprint.ts:26]: Thought it might be good to make the blueprint resemble our API reference and possibly list all methods under specific categories like we do in docs, e.g. category: Connect Webviews that groups all methods done by calling connect_webviews/*
Might be too much of a hassle though for not a lot of gain

File: src/lib/blueprint.ts (hunk 11 above)
Comment (kainpets) [reply src/lib/blueprint.ts:25]: Screenshot 2024-06-21 at 00 15 23
That's what I had in mind

Comment (razor-x) [reply src/lib/blueprint.ts:21]: How is this different than just the route description?

File: src/lib/blueprint.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (razor-x) [reply src/lib/blueprint.ts:22]: No, this is really important! But the category as a string is not enough. We should define them properly. 

This is similar to the `namespace` property [defined here](https://github.com/seamapi/javascript-http/blob/bea6d987a58597692dc6b3640c39f953a979f819/generate-routes.ts#L92) and related to the tags in the openapi spec.

Note that the blueprint does NOT have to be efficient for storage, it should be efficient for usage, so IMO duplicating data is ok, e.g., we can do

```suggestion
interface Route {
  ...
  category: Category
}

interface Category {
  name: string // Devices
  description: string // Manage devices with Seam
  path: string // /devices
}

interface Blueprint {
  routes: Route[]
  categories: Category[]
}
```

Alternatively, we can organize all the routes under categories,

```suggestion
interface Route {
    path: string // /devices/list
}

interface Category {
  name: string // Devices
  description: string // Manage devices with Seam
  path: string // /devices
  routes: Route[]
}

interface Blueprint {
  categories: Category[]
}
```

[the remark this answers is not in the document]
Comment (razor-x) [reply src/lib/blueprint.ts:22]: When doing codegen or docgen, do we always do effectively two loops? First over the category, second over the routes in the category?

[the remark this answers is not in the document]
Comment (razor-x) [reply src/lib/blueprint.ts:22]: @kainpets Moving this comment into this issue so it's easier to find in the future: https://github.com/seamapi/blueprint/issues/1

[the remark this answers is not in the document]
Comment (razor-x) [reply src/lib/blueprint.ts:22]: @kainpets Continuing this comment into this issue so it's easier to find in the future: https://github.com/seamapi/blueprint/issues/1

File: src/lib/openapi.ts [hunk 17]
```diff
@@ -1,5 +1,91 @@
+import { openapi } from '@seamapi/types/connect'
```
Comment (razor-x) [inline src/lib/openapi.ts:1]: We should not import `@seamapi/types/connect` here. We need the blueprint to define the (subset) of the openapi schema we will parse and then `@seamapi/types/connect` has to implement it.

One issues we have had with the existing openapi schema is when stuff was missing. We can either import an Openapi schmea typed defined in a 3rd party lib, or define a stricter openapi schema here (the proposed approach).

Comment (razor-x) [reply src/lib/openapi.ts:1]: This test is already checking the blueprint generation for `@seamapi/types/connect` https://github.com/seamapi/blueprint/blob/main/test/seam-blueprint.test.ts

Comment (razor-x) [reply src/lib/openapi.ts:1]: This fixture can be manually crafted to test all edge cases / parsing in a more controlled way: https://github.com/seamapi/blueprint/blob/main/test/fixtures/types/openapi.ts

Comment (kainpets) [reply src/lib/openapi.ts:1]: Oh, okay. So the types repo will be solely responsible for the generation? Guess I could move my efforts there once we define what we need.

Comment (razor-x) [reply src/lib/openapi.ts:1]: No, this repo will generate the blueprint from the types repo. [This test](https://github.com/seamapi/blueprint/blob/main/test/seam-blueprint.test.ts) show exactly how the sdk gen or doc gen code would use this project. So for example, 

```ts
import { openapi } from '@seamapi/types/connect'
import { createBlueprint } from '@seamapi/blueprint'

const generateSdk = async () => {
  const blueprint = createBlueprint({ openapi })
  await generateAndWriteRouteClasses(blueprint.routes)
  ...more codegen
}

File: src/lib/blueprint.ts (hunk 11 above)
Comment (kainpets) [reply src/lib/blueprint.ts:21]: It's not. I just couldn't find a good name :)

File: src/lib/openapi.ts (hunk 17 above)
Comment (kainpets) [reply src/lib/openapi.ts:1]: Oh, okay. So it's the createBlueprint that consumes the openapi schema and creates a reusable, more accessible format? I'll try to implement it then.

File: src/lib/blueprint.ts [hunk 18]
```diff
@@ -1,15 +1,90 @@
...
+  endpoints: Endpoint[]
+  subroutes: Route[] | null
+}
+
+interface Namespace {
+  name: string
+  path: string
+}
+
 export interface Blueprint {
```
Comment (razor-x) [inline src/lib/blueprint.ts:38]: When defining the types, it's much easier read this code if we start at the top and define the nodes below. E.g., instead of

	
```
	interface Parameter {}
	interface Endpoint {}
	interface Route {}
	interface Namespace {}
	interface Blueprint {}
	```

Prefer this order


```
	interface Blueprint {}
	interface Route {}
	interface Namespace {}
	interface Endpoint {}
	interface Parameter {}
```

File: src/lib/blueprint.ts [hunk 19]
```diff
@@ -1,15 +1,90 @@
...
+  parameters: Parameter[]
+  response: Response
+}
+
+interface Route {
+  name: string
+  path: string
+  namespace: Namespace | null
+  endpoints: Endpoint[]
+  subroutes: Route[] | null
```
Comment (razor-x) [inline src/lib/blueprint.ts:30]:
```suggestion
  subroutes: Route[]
```

File: src/lib/blueprint.ts [hunk 20]
```diff
@@ -1,15 +1,90 @@
...
+  description: string
+}
+
+type Method = 'GET' | 'POST' | 'PUT' | 'DELETE' | 'PATCH'
+
+interface Endpoint {
+  name: string
+  path: string
+  method: Method[]
+  semanticMethod?: string
```
Comment (razor-x) [inline src/lib/blueprint.ts:19]:
```suggestion
  semanticMethod: Method
```

File: src/lib/blueprint.ts [hunk 21]
```diff
@@ -1,15 +1,90 @@
...
+interface Response {
+  description: string
+}
+
+type Method = 'GET' | 'POST' | 'PUT' | 'DELETE' | 'PATCH'
+
+interface Endpoint {
+  name: string
+  path: string
+  method: Method[]
```
Comment (razor-x) [inline src/lib/blueprint.ts:18]:
```suggestion
  methods: Method[]
```

File: src/lib/blueprint.ts (hunk 20 above)
Comment (razor-x) [reply src/lib/blueprint.ts:19]: I'm realizing we need to encode the preferred method, e.g., we have [a lot of cases like this](https://github.com/seamapi/javascript-http/blob/bea6d987a58597692dc6b3640c39f953a979f819/test/seam/connect/seam-http-request.test.ts#L53-L73)

```
{
  path: '/devices/list',
  methods: ['GET', 'POST'],
  semanticMethod: 'GET',
  preferredMethod: 'POST'
}
```

Basically, we have to use POST because the API cannot support GET for all params yet.

File: src/lib/blueprint.ts [hunk 22]
```diff
@@ -1,15 +1,90 @@
...
+  semanticMethod?: string
+  routeDescription?: string | undefined
```
Comment (razor-x) [inline src/lib/blueprint.ts:20]:
```suggestion
  description: string | null
```

File: src/lib/blueprint.ts [hunk 23]
```diff
@@ -1,15 +1,90 @@
...
+  parameters: Parameter[]
+  response: Response
+}
+
+interface Route {
+  name: string
+  path: string
```
Comment (razor-x) [inline src/lib/blueprint.ts:27]:
```suggestion
  path: string
  description: string | null
```

File: src/lib/blueprint.ts [hunk 24]
```diff
@@ -1,15 +1,90 @@
...
+
+type Method = 'GET' | 'POST' | 'PUT' | 'DELETE' | 'PATCH'
+
+interface Endpoint {
+  name: string
+  path: string
+  method: Method[]
+  semanticMethod?: string
+  routeDescription?: string | undefined
+  parameters: Parameter[]
```
Comment (razor-x) [inline src/lib/blueprint.ts:21]: We still need to figure out where to set this from, but for now we can default `isDeprecated` to false.

We can default deprecationMessage to `''` when `isDeprecated` is `false`.

```suggestion
  isDeprecated: boolean
  deprecationMessage: string
  parameters: Parameter[]
```

File: src/lib/blueprint.ts [hunk 25]
```diff
@@ -1,15 +1,90 @@
 import type { Openapi } from './openapi.js'
 
+interface Parameter {
+  name: string
+  isRequired: boolean
```
Comment (razor-x) [inline src/lib/blueprint.ts:5]: We should be able to read these from the openapi spec via https://github.com/seamapi/nextlove/pull/133/files


```suggestion
  isRequired: boolean
  isDeprecated: boolean
  deprecationMessage: string
```

File: src/lib/blueprint.ts (hunk 20 above)
Comment (kainpets) [reply src/lib/blueprint.ts:19]: Does it make sense then to even mention other 'supported' methods? Maybe we should just stick to the `preferredMethod` for now?

File: src/lib/blueprint.ts (hunk 24 above)
Comment (kainpets) [reply src/lib/blueprint.ts:15]: The endpoint can be deprecated as well? Or did I misunderstand something?

File: src/lib/blueprint.ts (hunk 20 above)
Comment (razor-x) [reply src/lib/blueprint.ts]: We need both `semanticMethod` and `preferredMethod` since, taking the js sdk as an example, `semanticMethod` determines some of the type naming (body vs params) but `preferredMethod` is used to determine the actual method used.

Having `methods` is useful IMO for us internally to track progress, e.g., we can compare the three properties to see which endpoints are missing methods, etc.

File: src/lib/blueprint.ts [hunk 26]
```diff
@@ -2,14 +2,109 @@ import type { Openapi } from './openapi.js'
 
 export interface Blueprint {
   name: string
+  routes: Route[]
 }
 
+interface Route {
+  name: string
+  path: string
+  description: string | null
```
Comment (razor-x) [inline src/lib/blueprint.ts:11]: We should require all routes have a description and default to `''`.

```suggestion
  description: string
```

We could emit warnings to the console things that are missing this, e.g.,

```
if (description.trim().length === 0) console.warn(`... has an empty description`

File: src/lib/blueprint.ts [hunk 27]
```diff
@@ -2,14 +2,109 @@ import type { Openapi } from './openapi.js'
...
+  name: string
+  // TODO generate isRequired, isDeprecated, deprecationMessage from https://github.com/seamapi/nextlove/pull/133/files
+  isRequired: boolean
+  isDeprecated: boolean
+  deprecationMessage: string
+  description: string
+}
+
+interface Response {
+  description: string
```
Comment (razor-x) [inline src/lib/blueprint.ts:45]: I'm not sure if I already commented about this, but let's scope the rest of `Response` out of this PR. So this looks fine for now Implementing this will require introducing a concept of resources which will be it's own discussion.

File: src/lib/blueprint.ts [hunk 28]
```diff
@@ -2,14 +2,127 @@ import type { Openapi } from './openapi.js'
...
   name: string
+  routes: Route[]
 }
 
+interface Route {
+  name: string
+  path: string
+  // descriptions will default to an empty string and emit a warning, e.g.
+  // if (description.trim().length === 0) console.warn(`... has an empty description`)
+  description: string
```
Comment (razor-x) [inline src/lib/blueprint.ts:13]:
```suggestion
```

File: src/lib/blueprint.ts [hunk 29]
```diff
@@ -2,14 +2,127 @@ import type { Openapi } from './openapi.js'
...
+  namespace: Namespace | null
+  endpoints: Endpoint[]
+  subroutes: Route[]
+}
+
+interface Namespace {
+  name: string
+  description: string
```
Comment (razor-x) [inline src/lib/blueprint.ts:21]:
```suggestion
```

**Comment by razor-x** [comment]
Scope down this PR to return empty blueprint, only focus on the scheam.

File: src/lib/blueprint.ts [hunk 30]
```diff
@@ -2,14 +2,65 @@ import type { Openapi } from './openapi.js'
...
+  namespace: Namespace | null
+  endpoints: Endpoint[]
+  subroutes: Route[]
+}
+
+interface Namespace {
+  name: string
+  path: string
+}
```
Comment (kainpets) [inline src/lib/blueprint.ts:20]: It's not clear to me whether we want to keep Namespace. It's usefulness is not clear to me but maybe I'm misunderstanding something

Comment (razor-x) [reply src/lib/blueprint.ts:20]: We should keep namespace, even though it only applies to ACS at the moment. I think the usefulness will be more clear when trying to use the blueprint for sdk gen.

I'm not sure how to encode the namespace exists inside the openapi spec. We may need to remove `name` for now and say a namespace is something that only contains routes.

So logic is something like find all namespaces, then build all route

```
if namespace if path contains no endpoints
is route if path contains at least one endpoint
a route has namespace if route is under the path of a namespace
```

File: src/lib/blueprint.ts [hunk 31]
```diff
@@ -2,14 +2,65 @@ import type { Openapi } from './openapi.js'
 
 export interface Blueprint {
   name: string
+  routes: Route[]
 }
 
+interface Route {
+  name: string
+  path: string
+  description: string
```
Comment (razor-x) [inline src/lib/blueprint.ts:11]:
```suggestion
```

File: src/lib/blueprint.ts [hunk 32]
```diff
@@ -2,14 +2,65 @@ import type { Openapi } from './openapi.js'
 
 export interface Blueprint {
   name: string
+  routes: Route[]
 }
 
+interface Route {
+  name: string
```
Comment (razor-x) [inline src/lib/blueprint.ts:9]:
```suggestion
```

File: src/lib/blueprint.ts [hunk 33]
```diff
@@ -2,14 +2,65 @@ import type { Openapi } from './openapi.js'
...
+  path: string
+  description: string
+  namespace: Namespace | null
+  endpoints: Endpoint[]
+  subroutes: Route[]
+}
+
+interface Namespace {
+  name: string
```
Comment (razor-x) [inline src/lib/blueprint.ts:18]:
```suggestion
```

File: src/lib/blueprint.ts [hunk 34]
```diff
@@ -2,14 +2,65 @@ import type { Openapi } from './openapi.js'
...
+  deprecationMessage: string
+  parameters: Parameter[]
+  request: Request
+  response: Response
+}
+
+interface Parameter {
+  name: string
+  isRequired: boolean
+  isDeprecated: boolean
```
Comment (razor-x) [inline src/lib/blueprint.ts:39]:
```suggestion
  isUndocumented: boolean
  isDeprecated: boolean
```

File: src/lib/blueprint.ts [hunk 35]
```diff
@@ -2,14 +2,65 @@ import type { Openapi } from './openapi.js'
...
+}
+
+interface Endpoint {
+  name: string
+  path: string
+  methods: Method[]
+  semanticMethod: Method
+  preferredMethod: Method
+  description: string
+  isDeprecated: boolean
```
Comment (razor-x) [inline src/lib/blueprint.ts:29]:
```suggestion
  isUndocumented: boolean
  isDeprecated: boolean
```

**Review by razor-x — APPROVED** [review]

**Review by andrii-balitskyi — APPROVED** [review]
